Add ctrl-f/g support for notebooks and text files

aschlaep opened the pull request:
Fixes #1074 

### Goals
This is the start of a document search framework.  The goal is to provide default search functionality for notebooks and text documents, but provide the option to override the default behavior or add custom functionality for custom widgets.

### What this does
This PR creates the document search extension as a starting point for these search goals.  It provides a default implementation for searching notebook inputs (code and markdown) as well as text documents (opened in codemirror).  There are two search providers in this PR, one for codemirror and one for notebooks (which uses the codemirror search provider).  The default functionality provided includes searching forward and backward, counting matches, tracking match index, case sensitive search, and regex search.  I was able to make use of codemirror's searchcursor and overlay functionality, but had to mostly rewrite its search.js to provide my own hooks and match tracking.

Current behavior is that when starting a search, it'll look through all of your code and markdown cells for matches.  It'll unhide any hidden inputs containing matches and unrender any markdown containing matches.  It'll highlight all matches and put the cursor on the next match after the last cursor position in the active cell.  Once done searching, it'll re-render the markdown cells that were temporarily un-rendered (I chose not to re-hide code cells, but that could easily change).  The search can then toggle through matches across cells and loop at the beginning/end of the notebook.  If you move your cursor mid-search, it'll start toggling through matches from the new cursor position.  Thanks to codemirror's awesome overlay, the highlighting will update live as you type. 

Also in this PR, I removed the custom types living in the codemirror package in favor of those that exist in @types/codemirror.  I added back the types that were used by JupyterLab and do not exist in DefinitelyTyped's set of types.  I also had to expose several additional functions on codemirror to perform the search, get the cursor, etc.

### Future work
Potential (hopefully near-term) future enhancements to this would include search & replace, searching outputs, and searching across multiple files.

### To do:
- [x] replace UI used for testing with a proper overlay above active document
- [x] discuss interfaces
- [x] reorganize the commands
- [x] address case of adding/deleting cells mid-search
- [x] ensure search works with multiple documents open
- [x] more testing/bug squashing


Here's a gif demo with the UI I've been using for testing.
![ctrl-f-demo](https://user-images.githubusercontent.com/1820420/50369210-56958580-0546-11e9-8ca9-eb1556109e25.gif)

jasongrout commented:
@aschlaep - here is the recent work on React widgets: https://github.com/jupyterlab/jupyterlab/pull/5479

That PR is a bit overwhelming, but when you get a chance to push some of your react code up here, we can take a look at how to use this new react infrastructure.

aschlaep commented on packages/documentsearch-extension/src/searchbox.ts:
```diff
@@ -0,0 +1,242 @@
+    node.appendChild(search);
+    node.appendChild(results);
+
+    return node;
```
Yeah I'm currently working on an entire remake of the UI to replace this implementation with a more `React`ive implementation in a tsx.

aschlaep commented on packages/documentsearch-extension/src/searchbox.ts:
```diff
@@ -0,0 +1,242 @@
+    } catch (e) {
+      console.error('invalid regex: ', e);
+    }
+    if (ret.test('')) {
```
Yep good call, I'll fix that in the new UI

aschlaep commented on packages/documentsearch-extension/src/searchbox.ts:
```diff
@@ -0,0 +1,242 @@
+    const flag = caseSensitive ? 'g' : 'gi';
+    const queryText = regex
+      ? queryString
+      : queryString.replace(/[-[\]/{}()*+?.\\^$|]/g, '\\$&');
```
Will do!

aschlaep commented on packages/codemirror/typings/codemirror/codemirror.d.ts:
```diff
@@ -3,48 +3,23 @@
 // Definitions by: mihailik <https://github.com/mihailik>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
```
Hmm good call, the definitions in this file now are actually explicitly the ones that DefinitelyTyped *does not* cover, I'll update the comment.

aschlaep commented on packages/codemirror/src/editor.ts:
```diff
@@ -760,8 +821,10 @@ export class CodeMirrorEditor implements CodeEditor.IEditor {
+      anchor: this._toCodeMirrorPosition(range.start),
+      head: this._toCodeMirrorPosition(range.end),
+      from: this._toCodeMirrorPosition.bind(this, range.start),
+      to: this._toCodeMirrorPosition.bind(this, range.end)
```
For some reason, CodeMirror's Range type takes `anchor` and `head` as values and `from` and `to` as functions.  I'll have to double check my assumptions here about the values desired here.  There's a bit of annoying translation needed in some places between the CodeMirror Range type and the CodeEditor IRange.

jasongrout replied:
Actually, I was reading it wrong. I think you are eventually going to call this function: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/89592f7ec313174363136fe91749cedff272a70e/types/codemirror/index.d.ts#L294

That function needs a `{from: CodeMirror.Position, to: CodeMirror.Position}` range, not a `CodeMirror.Range` object. So I think we perhaps don't need this private function after all?

jasongrout replied:
Or at least, if we do want this private function, it is not returning a `CodeMirror.Range`, just a `{from: position, to: position}` object.

I can understand the confusion. The CodeMirror docs say it takes a range, but looking at the codemirror source, it's clear it wasn't talking about a `CodeMirror.Range` object, but rather a simple object with `from` and `to` positions (not functions).

jasongrout replied:
@aschlaep - I think this should work for now, and we can delete the extra argument to the scrollIntoView if we submit a change to the codemirror typings to make the margin argument optional in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ee0aad13bc3dd0f0176ca6794aacc0227d88c0fe/types/codemirror/index.d.ts#L286 and https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ee0aad13bc3dd0f0176ca6794aacc0227d88c0fe/types/codemirror/index.d.ts#L294

aschlaep commented:
Put together a UI that's closer to the final thing (need some proper up/down arrows, some minor alignment tweaks), here's a quick demo.
![ctrl-f-demo-2](https://user-images.githubusercontent.com/1820420/51067479-9b519100-15c7-11e9-9380-d93b77efd2c1.gif)

aschlaep commented:
This is now ready for review.  I'm still missing the resources for the buttons, but that should be a minor update (hopefully tomorrow).  Near-term future work for the document search extension include writing tests, focus tweaks (focus the document after closing search), perhaps switching the search box to a toolbar widget, and beginning work on search/replace and whole workspace search. @jasongrout

jasongrout commented on packages/codemirror/package.json:
```diff
@@ -42,11 +42,12 @@
     "@phosphor/disposable": "^1.1.2",
     "@phosphor/signaling": "^1.2.2",
     "@phosphor/widgets": "^1.6.0",
+    "@types/codemirror": "~0.0.70",
```
Does this belong here, or in the devDependencies below?

jasongrout commented on packages/codemirror/typings/codemirror/codemirror.d.ts:
```diff
@@ -100,891 +52,10 @@ declare namespace CodeMirror {
-
     /** Scrolls the given element into view. pos is a { from, to } object, in editor-local coordinates.
      The margin parameter is optional. When given, it indicates the amount of pixels around the given area that should be made visible as well. */
     scrollIntoView(pos: Range, margin?: number): void;
```
The typing for scrollIntoView should be deleted - I think the code does not take a Codemirror `Range` object, but rather the signature at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/89592f7ec313174363136fe91749cedff272a70e/types/codemirror/index.d.ts#L294 is what is needed.

jasongrout commented on packages/documentsearch-extension/src/index.ts:
```diff
@@ -0,0 +1,367 @@
+   * Initialize the search using the provided options.  Should update the UI
+   * to highlight all matches and "select" whatever the first match should be.
+   *
+   * @param options All of the search parameters configured in the search panel
```
There is no 'options' argument for this function. Can you instead explain `query` and `searchTarget`?

jasongrout commented on packages/documentsearch-extension/src/searchoverlay.tsx:
```diff
@@ -0,0 +1,286 @@
+
+function SearchIndices(props: ISearchIndexProps) {
+  return (
+    <>
```
Do you need a fragment here, where you're just returning one item?

aschlaep commented:
Just added icon resources and updated stylings for themes.  In the screenshots, case sensitivity is off, regex search is on.  There's an intermediate opacity level on hover as well that's not shown.
![screen shot 2019-01-28 at 3 02 27 pm](https://user-images.githubusercontent.com/1820420/51872263-a74e8a00-230d-11e9-8be7-69558aeb3a14.png)
![screen shot 2019-01-28 at 3 02 08 pm](https://user-images.githubusercontent.com/1820420/51872266-a9184d80-230d-11e9-8ce8-e557bc43db8a.png)

jasongrout commented:
Nice! Here's the layout I'm seeing in Chrome:

<img width="345" alt="screen shot 2019-01-28 at 3 19 49 pm" src="https://user-images.githubusercontent.com/192614/51872996-44aabd80-2310-11e9-9c38-c5fcf46db035.png">

and in Firefox:
<img width="359" alt="screen shot 2019-01-28 at 3 20 02 pm" src="https://user-images.githubusercontent.com/192614/51872999-48d6db00-2310-11e9-975d-029e5baa41f7.png">

Can you also see how things aren't lined up in Firefox?

jasongrout commented on packages/documentsearch-extension/src/searchoverlay.tsx:
```diff
@@ -0,0 +1,285 @@
+    this.state = props.overlayState;
+  }
+
+  private onChange(event: React.ChangeEvent) {
```
(for here and elsewhere): our convention is that private variables and methods start with an underscore, like `_onChange`.

jasongrout replied:
(this and other conventions are noted in the style guide: https://github.com/jupyterlab/jupyterlab/wiki/TypeScript-Style-Guide)

jasongrout commented:
Two minor keyboard issues:

1. In the notebook in *edit* mode, pressing accel-f opens the search overlay but does not focus it. Starting from command mode does focus the overlay.
2. When focused on the overlay, pressing escape does not exit the overlay.

jasongrout commented:
Andrew, I just merged in master and fixed the search extension to work with master. It was indeed a simple renaming.

jasongrout commented:
@aschlaep - ready for final review?

aschlaep commented:
Ready now!  Caught one more bug and made a styling update, it looks like something I merged in caused the button toggles to get box-sizing: unset, which caused them to get bigger and repeat their background, so I set the buttons to no background repeat as well as setting the input toggles to box-sizing: border-box, which seemed to fix the issue in both chrome and firefox.  The bug was an error that occurred when it tried to re-render a markdown cell containing a match after that cell had been deleted.

jasongrout commented:
Grant, if you want to pull this pr and test, that would be awesome

gnestor commented:
I tried checking out the branch and `jlpm`ing and running lab in dev-mode watch mode, but I don't see the _find_ left sidebar panel...

jasongrout commented:
Oh, it's better now! Ctrl f it use the command palette

gnestor commented:
Oh ya! First impression: the UX is nice!

cquah commented:
Thank you @aschlaep @jasongrout !!

aschlaep commented:
Awesome, thanks @jasongrout!
